jib-core/src/main/java/com/google/cloud/tools/jib/api/ImageReference.java

```diff
@@ -334,18 +360,27 @@ public String toString() {
     // Use tag if not the default tag.
     if (!DEFAULT_TAG.equals(tag)) {
       // Append with "@tag" instead of ":tag" if tag is a digest
-      referenceString.append(isTagDigest() ? '@' : ':').append(tag);
+      referenceString.append(':').append(tag);
+    }
+
+    if (!Strings.isNullOrEmpty(digest)) {
+      referenceString.append('@').append(digest);
     }
 
     return referenceString.toString();
   }
 
   /**
-   * Stringifies the {@link ImageReference}, without hiding the tag.
+   * Stringifies the {@link ImageReference}, including the default tag if no tag or digest is set.
    *
-   * @return the image reference in Docker-readable format, without hiding the tag
+   * @return the image reference in Docker-readable format, including the default tag if no tag or
+   *     digest is set.
    */
-  public String toStringWithTag() {
+  public String toStringWithQualifier() {
+    // Insert tag before digest
+    if (!Strings.isNullOrEmpty(digest)) {
```

Comment:
Hmm... so if both a digest and a non-default tag are given, now `toStringWithQualifier()` will print both. Are we good with that?

Reply:
Yes, I think this is correct.

Reply:
Just confirming: if so (e.g., returning `image:sometag@sha256...`), isn't it more natural for this method to consistently return `image:latest@sha256...` when the tag is `latest`? Currently, it returns `image@sha256...`. But if the intention is to make it return at least a tag or a digest, I can live with that. However, it's hard to predict the behavior just by reading the Javadoc.

Reply:
Yes, this is intended. Maybe always returning a tag would be more consistent, but again this is used in offline mode and we don't want to start returning `image:latest@sha256` or people will need to repopulate the cache.

`image:sometag@sha256` wasn't possible before this PR, so in this case we could go either way. Maybe digest should just take precedence over tag instead of returning both?

Reply:
> Maybe digest should just take precedence over tag instead of returning both?

I actually liked this. That is, this method appends a digest (`image@sha256`) if (and only if) it's set; otherwise, it appends a tag, including `latest`.

That said, I just realized some issue with our assumption with `latest`. It is indeed true that `latest` is assumed when no tag or digest is given. And so tar, we were okay, as it was not possible to have a digest and a tag at the same time. But what if the user simply referred to a base image only by a digest? In many cases, on an actual repository, the manifest would not be attached a `latest` tag. The question is then, would it make sense for `toString()` to append an unwarranted `:latest` that is possibly incorrect in practice? Perhaps if a digest is given but not a tag, we shouldn't assume the tag is `latest`?

Reply:
In that case should tag be `null` if a digest is set? Seems like it will need more changes outside of this method.

Reply:
Agreed. I'm more convinced that `latest` is the default value only when nothing is given. When a digest is given, it doesn't make sense to add `latest`.

Reply:
Ok. Do you think we should make tag `null` when the digest is set in this PR, or wait for a followup? I'm not yet sure how many extra changes making `tag` nullable might involve.

Reply:
If you are committed to fix all these in a quick succession, I am fine with a follow-up. At least it should be fixed before we can make a new release. Do you think this PR is ready for merging?

Reply:
I haven't checked [this](https://github.com/GoogleContainerTools/jib/pull/2381#discussion_r402538376) yet, but other than that, yeah I think it's ready.

Reply:
Alright. I think checking that also depends on how we assume `latest`, so we can check it in a follow-up.